subprojects/studio-plugin/src/main/java/org/gradle/profiler/studio/plugin/system/AndroidStudioSystemHelper.java

```diff
@@ -35,38 +43,54 @@ public static GradleSyncResult doGradleSync(Project project) {
      * Registers a listener that waits on next gradle sync if it's in progress.
      */
     public static void waitOnPreviousGradleSyncFinish(Project project) {
-        GradleProfilerGradleSyncListener syncListener = new GradleProfilerGradleSyncListener();
-        MessageBusConnection connection = GradleSyncState.subscribe(project, syncListener);
-        if (!GradleSyncState.getInstance(project).isSyncInProgress()) {
+        MessageBusConnection connection = project.getMessageBus().connect();
+        CompletableFuture<String> future = new CompletableFuture<>();
+        connection.subscribe(PROJECT_SYSTEM_SYNC_TOPIC, syncResult -> future.complete("Done"));
+        if (!ProjectSystemUtil.getSyncManager(project).isSyncInProgress()) {
             // Sync was actually not in progress,
             // just acknowledge the listener, so it won't wait forever.
-            syncListener.syncSkipped(project);
+            future.complete("Done");
         }
         try {
-            syncListener.waitAndGetResult();
+            future.join();
         } finally {
             connection.disconnect();
         }
     }
 
+    private static GradleProfilerGradleSyncListener subscribeToGradleSync(Project project) {
+        MessageBusConnection connection = project.getMessageBus().connect();
+        GradleProfilerGradleSyncListener syncListener = new GradleProfilerGradleSyncListener(connection);
+        connection.subscribe(GradleSyncState.GRADLE_SYNC_TOPIC, syncListener);
+        return syncListener;
+    }
+
     /**
      * Wait on Android Studio indexing and similar background tasks to finish.
-     *
+     * <p>
      * It seems there is no better way to do it atm.
      */
     public static void waitOnBackgroundProcessesFinish(Project project) {
         IdeFrame frame = WindowManagerEx.getInstanceEx().findFrameFor(project);
         StatusBarEx statusBar = frame == null ? null : (StatusBarEx) frame.getStatusBar();
         if (statusBar != null) {
-            statusBar.getBackgroundProcesses().forEach(it -> waitOn(it.getSecond()));
+            statusBar.getBackgroundProcesses().forEach(it -> waitOnProgressIndicator(it.getSecond()));
         }
     }
 
+    private static void waitOnProgressIndicator(ProgressIndicator progressIndicator) {
+        wait(progressIndicator::isRunning, WAIT_ON_PROCESS_SLEEP_TIME);
+    }
+
+    public static void waitOnPostStartupActivities(Project project) {
+        wait(() -> !StartupManager.getInstance(project).postStartupActivityPassed(), WAIT_ON_STARTUP_SLEEP_TIME);
+    }
+
     @SuppressWarnings("BusyWait")
-    private static void waitOn(ProgressIndicator progressIndicator) {
-        while (progressIndicator.isRunning()) {
+    private static void wait(Supplier<Boolean> whileCondition, long sleepMillis) {
```

Comment:
Nit: I think it results in somewhat nicer code at the call site if you put the lambda as the last parameter.

Reply:
Unfortunately, Java is not Groovy or Kotlin to have any advantage from that, but changed anyway